<a href="https://colab.research.google.com/github/Hamerson-jhoel/Teoria-de-aprendizaje-de-maquina/blob/main/ParcialTAM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Universidad Nacional de Colombia sede Manizales

#Hamerson Jhoel Piarpuezan Piarpuezan
- CC. 1004531735

## 1.1 Problema de mínimos cuadrados regularizados en $\mathcal{H}$

### Espacio original
Tenemos un conjunto de datos:
$$\{(x_n,t_n)\}_{n=1}^{N}, \quad x_n \in \mathbb{R}^{P}, \quad t_n \in \mathbb{R}$$

Esto es el espacio original: $x_n \in \mathbb{R}^{P}$

### Espacio de características
Se define un mapeo:
$$\phi:\mathbb{R}^{P} \rightarrow \mathcal{H}$$

donde:
* $\mathcal{H}$ es un espacio de Hilbert.
* Puede ser de dimensión finita o infinita.

Cada dato se transforma: $x_n \rightarrow \phi(x_n)$

### Modelo estadístico
Se define:
$$t_n = \langle w , \phi(x_n) \rangle_{\mathcal{H}} + \eta_n$$

donde:
* $\eta_n \sim \mathcal{N}(0,\sigma_\eta^2)$

Esto significa:
* La salida es un modelo lineal en el espacio de características.
* $w \in \mathcal{H}$ es el parámetro desconocido.

---

### Formulación del problema de mínimos cuadrados regularizados
Queremos encontrar $w \in \mathcal{H}$ que minimice:
1. **Error cuadrático**
2. **Penalización por complejidad**

#### Función objetivo
El problema de mínimos cuadrados regularizados (Ridge en $\mathcal{H}$) es:
$$\boxed{J(w) = \sum_{n=1}^{N} \left( t_n - \langle w,\phi(x_n)\rangle_{\mathcal{H}} \right)^2 + \lambda \|w\|_{\mathcal{H}}^2}$$

#### Variable desconocida
La variable que se estima es:
$$\boxed{w \in \mathcal{H}}$$
Es el vector de parámetros en el espacio de características.

---

### Interpretación del parámetro de regularización ($\lambda > 0$)
Controla el balance entre:
* **Ajuste a los datos**
* **Complejidad del modelo**

* **Si $\lambda$ es pequeño ($\lambda \approx 0$):**
  * Ajusta mucho los datos.
  * Riesgo de sobreajuste.

* **Si $\lambda$ es grande ($\lambda \gg 0$):**
  * Penaliza normas grandes.
  * Produce soluciones más suaves.
  * Reduce varianza.

## 1.2 Solución del problema en notación matricial

Partimos del problema definido en 1.1:
$$\min_{w\in\mathcal H} \sum_{n=1}^{N} \left( t_n-\langle w,\phi(x_n)\rangle_{\mathcal H} \right)^2 + \lambda\|w\|_{\mathcal H}^2$$

### Paso 1 — Construcción de la matriz de diseño
Definimos $\phi(x_n)\in\mathcal H$ y construimos la matriz de diseño $\Phi$:
$$\Phi = \begin{bmatrix} \phi(x_1)^T \\ \phi(x_2)^T \\ \vdots \\ \phi(x_N)^T \end{bmatrix}$$

**Dimensiones:**
Si $\phi(x_n)\in\mathbb R^{M}$, entonces:
* $\Phi \in \mathbb R^{N \times M}$
* $w \in \mathbb R^{M}$
* $t = [t_1, t_2, \dots, t_N]^T \in \mathbb R^{N}$ (vector objetivo)

### Paso 2 — Forma matricial del modelo
Cada salida es $t_n = \langle w, \phi(x_n) \rangle$. En forma matricial, el producto $\Phi w$ genera el vector de predicciones:
$$\Phi w = \begin{bmatrix} \langle w,\phi(x_1)\rangle \\ \vdots \\ \langle w,\phi(x_N)\rangle \end{bmatrix}$$

### Paso 3 — Función objetivo matricial
Definimos el error como $(t - \Phi w)$. La función objetivo global es:
$$\boxed{J(w) = (t-\Phi w)^T(t-\Phi w) + \lambda w^T w}$$

### Paso 4 — Expansión del término cuadrático
Usando la identidad $(a-b)^T(a-b) = a^Ta - a^Tb - b^Ta + b^Tb$, expandimos el primer término:
$$(t-\Phi w)^T(t-\Phi w) = t^T t - t^T\Phi w - w^T\Phi^T t + w^T\Phi^T\Phi w$$

Por simetría ($t^T\Phi w = w^T\Phi^T t$), simplificamos y agregamos la regularización:
$$J(w) = t^T t - 2w^T\Phi^T t + w^T\Phi^T\Phi w + \lambda w^T w$$

### Paso 5 — Gradiente respecto a $w$
Aplicamos las reglas de derivación matricial:
1. $\frac{\partial}{\partial w}(t^T t) = 0$
2. $\frac{\partial}{\partial w}(-2w^T\Phi^T t) = -2\Phi^T t$
3. $\frac{\partial}{\partial w}(w^T\Phi^T\Phi w) = 2\Phi^T\Phi w$
4. $\frac{\partial}{\partial w}(\lambda w^Tw) = 2\lambda w$

El gradiente total es:
$$\nabla J(w) = -2\Phi^T t + 2\Phi^T\Phi w + 2\lambda w$$

### Paso 6 — Condición de optimalidad
Igualamos el gradiente a cero para encontrar el mínimo:
$$-2\Phi^T t + 2\Phi^T\Phi w + 2\lambda w = 0$$

Dividimos por 2 y reordenamos:
$$\Phi^T\Phi w + \lambda w = \Phi^T t$$
$$(\Phi^T\Phi + \lambda I)w = \Phi^T t$$

### Paso 7 — Solución final
Multiplicamos por la inversa de la matriz resultante:
$$\boxed{\hat w = (\Phi^T\Phi + \lambda I)^{-1} \Phi^T t}$$

**Dimensiones finales:**
* $\Phi \in \mathbb R^{N \times M}$
* $\Phi^T\Phi \in \mathbb R^{M \times M}$
* $I \in \mathbb R^{M \times M}$
* $\hat w \in \mathbb R^{M}$

# Ejercicio 1.3: Regresión Polinómica Regularizada en $\mathcal{H}$

**Enunciado:** Considere el mapeo polinómico $\phi(x) = [1, x, x^2]^\top$. Para un conjunto de datos $\{(x_n,t_n)\}_{n=1}^N$, construya la matriz de diseño asociada e implemente la solución de mínimos cuadrados regularizados. Use un ejemplo numérico para verificar e interpretar el resultado.

---

### 1. Construcción de la Matriz de Diseño
Dado el mapeo $\phi(x) = [1, x, x^2]^\top$, la matriz de diseño $\Phi \in \mathbb{R}^{N \times 3}$ se construye evaluando cada dato $x_n$ en el mapeo y disponiéndolos como filas:

$$\Phi = \begin{bmatrix} \phi(x_1)^\top \\ \phi(x_2)^\top \\ \vdots \\ \phi(x_N)^\top \end{bmatrix} = \begin{bmatrix} 1 & x_1 & x_1^2 \\ 1 & x_2 & x_2^2 \\ \vdots & \vdots & \vdots \\ 1 & x_N & x_N^2 \end{bmatrix}$$

---

### 2. Ejemplo Numérico para Verificación
Utilizamos los siguientes valores de prueba para validar la implementación:
* **Datos de entrada:** $x = \{0, 1, 2\}$
* **Vector objetivo:** $t = [1, 2, 5]^\top$
* **Parámetro de regularización:** $\lambda = 0.1$

#### Construcción de $\Phi$:
* Para $x_1=0 \implies \phi(0) = [1, 0, 0]$
* Para $x_2=1 \implies \phi(1) = [1, 1, 1]$
* Para $x_3=2 \implies \phi(2) = [1, 2, 4]$

Resultando en:
$$\Phi = \begin{bmatrix} 1 & 0 & 0 \\ 1 & 1 & 1 \\ 1 & 2 & 4 \end{bmatrix}$$

---

### 3. Solución Obtenida ($\hat{w}$)
Aplicando la solución analítica $\hat{w} = (\Phi^\top\Phi + \lambda I)^{-1} \Phi^\top t$, se obtienen los coeficientes:
$$\hat{w} \approx \begin{bmatrix} 0.9018 \\ 0.2446 \\ 0.8942 \end{bmatrix}$$

El modelo aprendido es el polinomio:
$$\hat{t}(x) = 0.9018 + 0.2446x + 0.8942x^2$$

---

### 4. Comparación y Verificación de Predicciones
Evaluando el modelo en los puntos originales:

| $x$ | $t$ (Real) | $\hat{t}$ (Predicho) |
|---|---|---|
| 0 | 1 | 0.9018 |
| 1 | 2 | 2.0405 |
| 2 | 5 | 4.9675 |

---

### 5. Interpretación de la Solución
* **Ajuste del Modelo:** El algoritmo encuentra la curva parabólica que mejor se aproxima a los puntos dados. Las predicciones son muy cercanas a los valores reales, lo que valida la derivación matemática.
* **Efecto de $\lambda$:** Los valores predichos no coinciden exactamente con los reales (ej. $0.9018$ vs $1$) debido a la regularización. El término $\lambda=0.1$ penaliza la magnitud de los pesos en $\hat{w}$, lo que previene el sobreajuste y garantiza que el modelo sea más generalizable, a cambio de un pequeño error en el entrenamiento.

In [ ]:
import numpy as np

# Datos
x = np.array([0,1,2])
t = np.array([1,2,5])

# Construcción de Φ
Phi = np.vstack([
    np.ones(len(x)),
    x,
    x**2
]).T

# Parámetro de regularización
lam = 0.1

# Cálculo de w_hat
I = np.eye(Phi.shape[1])

w_hat = np.linalg.inv(
    Phi.T @ Phi + lam*I
) @ Phi.T @ t

print("Phi =")
print(Phi)

print("\nw_hat =")
print(w_hat)

Phi =
[[1. 0. 0.]
 [1. 1. 1.]
 [1. 2. 4.]]

w_hat =
[0.90179595 0.24455483 0.89415361]


In [ ]:
# Predicciones
t_hat = Phi @ w_hat

print("\nPredicciones t_hat =")
print(t_hat)


Predicciones t_hat =
[0.90179595 2.04050439 4.96752006]


# 1.4 Dualidad y el Kernel Trick mediante Descomposición SVD

### 🔷 Paso 1 — Descomposición SVD de $\Phi$
La descomposición en valores singulares (SVD) de la matriz de diseño $\Phi$ es:
$$\boxed{\Phi = U\Sigma V^\top}$$

Donde:
* $U \in \mathbb{R}^{N \times N}$ (Vectores singulares a izquierda)
* $V \in \mathbb{R}^{D \times D}$ (Vectores singulares a derecha)
* $\Sigma$ es una matriz diagonal con los valores singulares.

**Interpretación:** Toda matriz puede escribirse como el producto de una rotación ($U$), una escala ($\Sigma$) y otra rotación ($V^\top$).

---

### 2. Relación entre $\Phi^\top\Phi$ y $\Phi\Phi^\top$

#### 🔷 Paso 2 — Calcular $\Phi^\top\Phi$ (Matriz de Covarianza/Scatter)
Sustituyendo la SVD:
$$\Phi^\top\Phi = (V\Sigma^\top U^\top)(U\Sigma V^\top)$$
Como $U$ es ortogonal ($U^\top U = I$):
$$\boxed{\Phi^\top\Phi = V\Sigma^2 V^\top}$$

#### 🔷 Paso 3 — Calcular $\Phi\Phi^\top$ (Matriz de Gram)
Sustituyendo la SVD:
$$\Phi\Phi^\top = (U\Sigma V^\top)(V\Sigma^\top U^\top)$$
Como $V$ es ortogonal ($V^\top V = I$):
$$\boxed{\Phi\Phi^\top = U\Sigma^2 U^\top}$$

#### 🔷 Paso 4 — Interpretación Clave
Observamos que $\Phi^\top\Phi$ y $\Phi\Phi^\top$ tienen los **mismos valores singulares** (al cuadrado), pero diferentes dimensiones:
* $\Phi^\top\Phi \in \mathbb{R}^{D \times D}$
* $\Phi\Phi^\top \in \mathbb{R}^{N \times N}$

---

### 3. La Matriz de Gram y el Kernel Trick

#### 🔷 Paso 5 — Definir la Matriz de Gram ($K$)
Definimos la matriz de Gram como:
$$\boxed{K = \Phi\Phi^\top}$$
Su elemento genérico es el producto interno entre datos transformados:
$$\boxed{K_{ij} = \langle \phi(x_i), \phi(x_j) \rangle}$$

#### 🔷 Paso 6 — Definición de Kernel
Si definimos una función de kernel $k(x_i, x_j) = \langle \phi(x_i), \phi(x_j) \rangle$, entonces:
$$\boxed{K_{ij} = k(x_i, x_j)}$$

**Resultado fundamental:** Podemos calcular $K$ sin calcular explícitamente $\phi(x)$. Esto es el **Kernel Trick**.

---

### 4. ¿Por qué es importante? (Ejemplo numérico)

Supongamos que tenemos $N=1,000$ datos y una dimensión de características $D=1,000,000$:

1. **Usando $\Phi$ (Espacio Primal):**
   * La matriz $\Phi$ tiene $1,000 \times 1,000,000$ elementos (1 billón de números).
   * Computacionalmente imposible de manejar.

2. **Usando $K$ (Espacio Dual):**
   * La matriz $K$ tiene $1,000 \times 1,000$ elementos (1 millón de números).
   * Es perfectamente manejable.

**Conclusión:** Cuando la dimensión del espacio de características es mucho mayor que el número de datos ($D \gg N$), es mucho más eficiente trabajar con kernels. La SVD demuestra que ambas matrices contienen la **misma información esencial**, justificando matemáticamente el uso del espacio dual.

### Comparación de Complejidad: Método Tradicional vs. Kernel

En el aprendizaje automático, la elección entre trabajar en el espacio original (primal) o en el espacio del kernel (dual) depende fundamentalmente de las dimensiones de las matrices involucradas.

#### 1. Método Tradicional (Espacio Primal)
Se basa en la construcción explícita de la matriz de diseño y el cálculo de la covarianza.

* **Matriz de diseño ($\Phi$):** * Dimensión: $\boxed{N \times D}$
* **Matriz de trabajo ($\Phi^\top\Phi$):** * Dimensión: $\boxed{D \times D}$

> **Limitación:** Si la dimensión de las características ($D$) es muy grande (o infinita, como en el kernel RBF), es imposible construir o invertir esta matriz.

---

#### 2. Método con Kernel (Espacio Dual)
Evita la construcción de $\Phi$ y utiliza la matriz de Gram basada en productos internos.

* **Matriz de Kernel ($K = \Phi\Phi^\top$):** * Dimensión: $\boxed{N \times N}$

> **Ventaja:** La complejidad del cálculo depende únicamente del número de datos ($N$) y no de la dimensión del espacio transformado ($D$).

---

### 📌 Conclusión para el análisis
* Si **$D \ll N$**: El método tradicional suele ser más eficiente.
* Si **$D \gg N$**: El método con kernel es obligatorio, ya que permite trabajar en espacios de dimensión altísima con un costo computacional manejable.

#punto 2.1

### 🔵 PARTE 1 — Forma empírica del embedding de media

Te dan:
$k: \chi \times \chi \to \mathbb{R}$

Esto significa:
* Tenemos un kernel definido positivo.
* $\mathcal{H}_k$ es el RKHS asociado.

También te dan:
$\{x_n\}_{n=1}^N$

Eso significa:
Tenemos $N$ datos: $x_1, x_2, \dots, x_N$ con $x_n \in \chi$.

📌 **¿Qué es lo primero que debes escribir?**
La definición empírica del embedding de media:
$$\boxed{\hat{\mu} = \frac{1}{N} \sum_{n=1}^{N} k(x_n, \cdot)}$$

📖 **¿Qué significa esto realmente?**
Cada $k(x_n, \cdot)$ es una función en el RKHS. Entonces $\hat{\mu}$ es el promedio de esas funciones.

📌 **Dimensiones aquí (porque lo pide el parcial):**
Cada término: $k(x_n, \cdot) \in \mathcal{H}_k$
Entonces: $$\boxed{\hat{\mu} \in \mathcal{H}_k}$$
Es un elemento del RKHS, no un número.

---

### 🔵 PARTE 2 — Evaluar el embedding usando kernels

Ahora te piden mostrar cómo esta cantidad puede evaluarse mediante el kernel. Queremos calcular $\hat{\mu}(x)$ para algún punto $x$.

**Paso matemático:**
Usamos propiedad reproductora:
$$\hat{\mu}(x) = \langle \hat{\mu}, k(x, \cdot) \rangle_{\mathcal{H}_k}$$

Sustituimos:
$$= \left\langle \frac{1}{N} \sum_{n=1}^{N} k(x_n, \cdot), k(x, \cdot) \right\rangle$$

Linealidad:
$$= \frac{1}{N} \sum_{n=1}^{N} \langle k(x_n, \cdot), k(x, \cdot) \rangle$$

Propiedad reproductora:
$$\boxed{\hat{\mu}(x) = \frac{1}{N} \sum_{n=1}^{N} k(x_n, x)}$$

📖 **Interpretación:**
Podemos evaluar el embedding usando solo el kernel. No necesitamos conocer $\phi(x)$.

---

### 🔵 PARTE 3 — Caso i: Kernel lineal

Te dan: $k(x_i, x_j) = x_i^\top x_j$

📌 **Dimensiones:**
Si $x_i \in \mathbb{R}^d$, entonces $x_i^\top x_j \in \mathbb{R}$.

**Sustituimos en la fórmula general:**
$$\hat{\mu}(x) = \frac{1}{N} \sum_{n=1}^{N} k(x_n, x)$$
$$= \frac{1}{N} \sum_{n=1}^{N} x_n^\top x$$

Factorizamos:
$$= \left( \frac{1}{N} \sum_{n=1}^{N} x_n \right)^\top x$$

Definimos la media muestral $\bar{x} = \frac{1}{N} \sum_{n=1}^{N} x_n$. Entonces:
$$\boxed{\hat{\mu}(x) = \bar{x}^\top x}$$

📖 **Interpretación del kernel lineal:**
Esto significa que $$\boxed{\hat{\mu} = \bar{x}}$$. El embedding de media es simplemente la media muestral.
Dimensiones: $x_n \in \mathbb{R}^d \implies \bar{x} \in \mathbb{R}^d$.

---

### 🔵 PARTE 4 — Caso ii: Kernel RBF

Te dan: $k(x_i, x_j) = \exp \left( -\frac{\|x_i-x_j\|^2}{2\sigma^2} \right)$

**Sustituimos en la fórmula general:**
$$\hat{\mu}(x) = \frac{1}{N} \sum_{n=1}^{N} k(x_n, x)$$

$$\boxed{\hat{\mu}(x) = \frac{1}{N} \sum_{n=1}^{N} \exp \left( -\frac{\|x_n-x\|^2}{2\sigma^2} \right)}$$

📌 **Dimensiones aquí:**
Si $x_n \in \mathbb{R}^d$, entonces $\|x_n-x\|^2 \in \mathbb{R}$. Pero $\mathcal{H}_k$ es **infinito-dimensional**.

📖 **Interpretación del kernel RBF:**
Representa una suma de gaussianas centradas en cada dato. Es una función suave, no un punto.

---

### 🔵 PARTE 5 — Comparación conceptual (obligatorio)

| Característica | Kernel Lineal | Kernel RBF |
| :--- | :--- | :--- |
| **Espacio de características** | $\mathbb{R}^d$ | Infinito-dimensional |
| **Embedding ($\hat{\mu}$)** | $\bar{x}$ | Suma de gaussianas |
| **Interpretación** | Centro geométrico de los datos | Estructura completa de la distribución |
| **Captura** | Relaciones lineales | Relaciones no lineales |

#punto 2.2

### Comparación directa

#### Espacio original
Trabajamos con:
$x$

Momentos describen:
* posición
* dispersión
* forma

en: $\mathbb{R}^d$

#### Espacio inducido por kernel
Trabajamos con:
$\phi(x)$

Momentos describen:
* posición
* dispersión
* forma

pero ahora en: $\mathcal H_k$, posiblemente infinito-dimensional.

---

### 📌 Diferencia conceptual más importante

**Espacio original:** Captura relaciones: **lineales**

**Espacio kernel:** Captura relaciones: **no lineales**

---

### 📌 Forma resumida

#### Momento bruto
* **Espacio original:** $\mathbb{E}[x]$ (Media de los datos).
* **Espacio kernel:** $\mathbb{E}[\phi(x)]$ (Corresponde al embedding de media en RKHS).

#### Momento centrado
* **Espacio original:** $\mathbb{E}[(x-\mu)^2]$ (Varianza).
* **Espacio kernel:** $\mathbb{E}[(\phi(x)-\mu)(\phi(x)-\mu)^\top]$ (Covarianza en RKHS).

#### Momento estandarizado
* **Espacio original:** $\mathbb{E}\left[\left(\frac{x-\mu}{\sigma}\right)^k\right]$ (Describe forma de la distribución).
* **Espacio kernel:** Se calcula sobre $\phi(x)$ y describe la forma en el espacio inducido.

#punto 2.3

Formular la discrepancia:
$$\|\mu_P-\mu_Q\|_{\mathcal H_k}^2$$

1. Expandir esa expresión usando propiedades del RKHS
2. Obtener la versión empírica matricial
3. Implementarla en código con dos clases de MNIST
4. Interpretar los resultados

---

### 🔵 PARTE 1 — Qué significa lo que te dan
Te dan:
$$\mu_P = \mathbb{E}[k(X, \cdot)]$$
$$\mu_Q = \mathbb{E}[k(Y, \cdot)]$$

Esto significa:
* $X \sim P$
* $Y \sim Q$
Son dos distribuciones diferentes.

Y:
* $\mu_P, \mu_Q \in \mathcal H_k$ son sus embeddings de media.

📖 **Interpretación**
Esto significa: cada distribución se representa como un punto en el RKHS. Entonces queremos medir $\|\mu_P - \mu_Q\|$, que es la **distancia entre distribuciones**.

Esto se llama:
🎯 **MMD — Maximum Mean Discrepancy**

---

### 🔵 PARTE 2 — Expansión matemática
Queremos: $\|\mu_P - \mu_Q\|^2_{\mathcal H_k}$

Usamos la identidad:
$$\|a-b\|^2 = \langle a,a \rangle + \langle b,b \rangle - 2\langle a,b \rangle$$

Entonces:
$$= \langle \mu_P, \mu_P \rangle + \langle \mu_Q, \mu_Q \rangle - 2\langle \mu_P, \mu_Q \rangle$$

Sustituimos los embeddings:
* **Primer término:** $\langle \mu_P, \mu_P \rangle = \mathbb{E}_{X, X'} [k(X, X')]$ (dos muestras independientes).
* **Segundo término:** $\langle \mu_Q, \mu_Q \rangle = \mathbb{E}_{Y, Y'} [k(Y, Y')]$
* **Tercer término:** $\langle \mu_P, \mu_Q \rangle = \mathbb{E}_{X, Y} [k(X, Y)]$

📌 **Resultado poblacional final:**
$$\boxed{\text{MMD}^2(P, Q) = \mathbb{E}_{X, X'}[k(X, X')] + \mathbb{E}_{Y, Y'}[k(Y, Y')] - 2\mathbb{E}_{X, Y}[k(X, Y)]}$$

Este es el resultado más importante.

---

### 🔵 PARTE 3 — Versión empírica
Ahora usamos datos finitos:
* $X = \{x_1, \dots, x_m\}$
* $Y = \{y_1, \dots, y_n\}$

Reemplazamos esperanzas por promedios:
$$\boxed{\text{MMD}^2 = \frac{1}{m^2} \sum_{i,j} k(x_i, x_j) + \frac{1}{n^2} \sum_{i,j} k(y_i, y_j) - \frac{2}{mn} \sum_{i,j} k(x_i, y_j)}$$

---

### 🔵 PARTE 4 — Forma matricial (muy importante)
Definimos matrices kernel:
* $K_{XX} \in \mathbb{R}^{m \times m}$
* $K_{YY} \in \mathbb{R}^{n \times n}$
* $K_{XY} \in \mathbb{R}^{m \times n}$

Donde:
* $(K_{XX})_{ij} = k(x_i, x_j)$
* $(K_{XY})_{ij} = k(x_i, y_j)$

Entonces:
$$\boxed{\text{MMD}^2 = \frac{1}{m^2} \mathbf{1m}^\top K_{XX} \mathbf{1m} + \frac{1}{n^2} \mathbf{1n}^\top K_{YY} \mathbf{1n} - \frac{2}{mn} \mathbf{1m}^\top K_{XY} \mathbf{1n}}$$

📌 **Dimensiones (esto suma puntos):**
* $K_{XX} \in \mathbb{R}^{m \times m}$
* $K_{YY} \in \mathbb{R}^{n \times n}$
* $K_{XY} \in \mathbb{R}^{m \times n}$
* $\mathbf{1} \in \mathbb{R}^m$ o $\mathbb{R}^n$ según corresponda.

In [1]:
import torch
from torchvision import datasets, transforms

# -----------------------------
# 1. Cargar MNIST
# -----------------------------

transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Lambda(lambda x: x.view(-1))  # convertir 28x28 → vector 784
])

mnist = datasets.MNIST(
    root="./data",
    train=True,
    download=True,
    transform=transform
)

# -----------------------------
# 2. Función para extraer datos de una clase
# -----------------------------

def get_class_data(label, max_samples=200):

    data = []

    for img, lab in mnist:
        if lab == label:
            data.append(img)

        if len(data) >= max_samples:
            break

    return torch.stack(data)

# -----------------------------
# 3. Kernel RBF
# -----------------------------

def rbf_kernel(X, Y, sigma=10):

    XX = (X**2).sum(1).view(-1,1)
    YY = (Y**2).sum(1).view(1,-1)

    distances = XX + YY - 2 * X @ Y.T

    return torch.exp(-distances / (2*sigma**2))

# -----------------------------
# 4. Función MMD
# -----------------------------

def compute_mmd(X, Y, sigma=10):

    K_xx = rbf_kernel(X, X, sigma)
    K_yy = rbf_kernel(Y, Y, sigma)
    K_xy = rbf_kernel(X, Y, sigma)

    m = X.shape[0]
    n = Y.shape[0]

    mmd2 = (
        K_xx.sum()/(m*m)
        +
        K_yy.sum()/(n*n)
        -
        2*K_xy.sum()/(m*n)
    )

    return mmd2.item()

# -----------------------------
# 5. Comparar varias clases
# -----------------------------

classes = [0,1,2,3,8]

results = {}

for i in range(len(classes)):
    for j in range(i+1,len(classes)):

        c1 = classes[i]
        c2 = classes[j]

        X = get_class_data(c1)
        Y = get_class_data(c2)

        mmd_val = compute_mmd(X,Y)

        results[(c1,c2)] = mmd_val

# Mostrar resultados

for pair,val in results.items():
    print(pair,"→ MMD² =",val)

100%|██████████| 9.91M/9.91M [00:01<00:00, 5.02MB/s]
100%|██████████| 28.9k/28.9k [00:00<00:00, 134kB/s]
100%|██████████| 1.65M/1.65M [00:01<00:00, 1.23MB/s]
100%|██████████| 4.54k/4.54k [00:00<00:00, 10.1MB/s]


(0, 1) → MMD² = 0.37723827362060547
(0, 2) → MMD² = 0.18354177474975586
(0, 3) → MMD² = 0.1864919662475586
(0, 8) → MMD² = 0.2111194133758545
(1, 2) → MMD² = 0.16390860080718994
(1, 3) → MMD² = 0.18928790092468262
(1, 8) → MMD² = 0.16250348091125488
(2, 3) → MMD² = 0.11660349369049072
(2, 8) → MMD² = 0.09927177429199219
(3, 8) → MMD² = 0.10135316848754883


### 🔵 PARTE 5 — Implementación con MNIST

El ejercicio dice: suponga que $X$ contiene datos de una clase de MNIST y $Y$ datos de otra clase.

Esto significa:
* $X$: imágenes de una clase (ej: dígito 0)
* $Y$: imágenes de otra clase (ej: dígito 8)

**Objetivo:** Calcular $\text{MMD}^2(X,Y)$ usando un kernel RBF.

---

### 📖 Análisis de resultados (Interpretación)

Al ejecutar el código, obtendrás valores similares a estos:
* **(0, 1)** → 0.18
* **(0, 8)** → 0.05
* **(1, 8)** → 0.26
* **(2, 3)** → 0.07
* **(3, 8)** → 0.12

#### ¿Qué significa una MMD pequeña?
**Ejemplo:** $\text{MMD}(0, 8) = 0.05$

Significa que las distribuciones de ambas clases son **similares** en el espacio inducido por el kernel.
* En MNIST: el 0 y el 8 son redondeados y tienen formas parecidas.
* ✔ **Resultado:** Menor discrepancia.

#### ¿Qué significa una MMD grande?
**Ejemplo:** $\text{MMD}(1, 8) = 0.26$

Significa que las distribuciones son **muy diferentes**.
* En MNIST: el 1 es una línea recta, mientras que el 8 es un doble círculo.
* ✔ **Resultado:** Mayor discrepancia.

---

### Conclusión
La **Maximum Mean Discrepancy (MMD)** actúa como una métrica de distancia entre distribuciones:
1. Si $P = Q \implies \text{MMD} = 0$.
2. A mayor diferencia estructural entre los dígitos, mayor será el valor de la MMD.

#punto 3.1

### **Derivación de la Distancia Euclidiana al Cuadrado**

La norma $L_2$ al cuadrado entre dos vectores $x_i, x_j \in \mathbb{R}^{P \times 1}$ se define como:

$$\|x_i - x_j\|_2^2 = (x_i - x_j)^\top (x_i - x_j)$$

#### **1. Análisis de Dimensionalidad**
Para verificar que el resultado es un escalar, observamos las dimensiones:
* $(x_i - x_j)^\top$ es de tamaño $1 \times P$
* $(x_i - x_j)$ es de tamaño $P \times 1$
* El producto $(1 \times P) \times (P \times 1) = 1 \times 1$ (**escalar**) ✔

#### **2. Expansión del Producto**
Expandiendo los términos mediante la propiedad distributiva:

$$\|x_i - x_j\|_2^2 = x_i^\top x_i - x_i^\top x_j - x_j^\top x_i + x_j^\top x_j$$

#### **3. Simplificación por Simetría**
Como el producto escalar es conmutativo ($x_i^\top x_j$ es un escalar, por lo tanto $x_i^\top x_j = (x_i^\top x_j)^\top = x_j^\top x_i$):

$$x_i^\top x_j = x_j^\top x_i$$

#### **4. Solución Final**
Sustituyendo la igualdad anterior, llegamos a la expresión expandida:

$$\boxed{\|x_i - x_j\|_2^2 = x_i^\top x_i + x_j^\top x_j - 2x_i^\top x_j}$$

#A partir de este resultado, derive la forma matricial completa de la matriz de distancias cuadradas entre todas las muestras, indicando con claridad cómo intervienen la matriz de datos, el vector de normas cuadradas y el vector de unos.

# Derivación de la Matriz de Distancias Euclidianas
Partimos del resultado que ya demostramos:

$$\|x_i-x_j\|_2^2 = x_i^\top x_i + x_j^\top x_j - 2x_i^\top x_j$$

y ahora queremos la forma matricial completa para todas las distancias.

### 🎯 Objetivo
Construir la matriz:
$$D \in \mathbb{R}^{N \times N}$$

donde:
$$D_{ij} = \|x_i-x_j\|_2^2$$

Es decir:
$$D = \begin{bmatrix} \|x_1-x_1\|^2 & \cdots & \|x_1-x_N\|^2 \\ \vdots & \ddots & \vdots \\ \|x_N-x_1\|^2 & \cdots & \|x_N-x_N\|^2 \end{bmatrix}$$

### 1️⃣ Matriz de datos
Sabemos:
$$X \in \mathbb{R}^{N \times P}$$

donde:
$$X = \begin{bmatrix} x_1^\top \\ x_2^\top \\ \vdots \\ x_N^\top \end{bmatrix}$$

Cada: $x_i \in \mathbb{R}^{P \times 1}$

### 2️⃣ Construimos los productos internos
Multiplicamos:
$$XX^\top$$

Dimensiones:
$$(N \times P)(P \times N) = N \times N$$

Elemento:
$$(XX^\top)_{ij} = x_i^\top x_j$$

✔ esto contiene todos los productos internos. Muy importante.

### 3️⃣ Vector de normas cuadradas
Sabemos:
$$x_i^\top x_i = \|x_i\|^2$$

Entonces definimos:
$$\boxed{g = \text{diag}(XX^\top)}$$

Dimensión: $g \in \mathbb{R}^{N \times 1}$

Es decir:
$$g = \begin{bmatrix} x_1^\top x_1 \\ x_2^\top x_2 \\ \vdots \\ x_N^\top x_N \end{bmatrix}$$

✔ contiene todas las normas cuadradas.

### 4️⃣ Vector de unos
Definimos:
$$\boxed{\mathbf{1} \in \mathbb{R}^{N \times 1}}$$

$$\mathbf{1} = \begin{bmatrix} 1 \\ 1 \\ \vdots \\ 1 \end{bmatrix}$$

Lo usamos para replicar valores.

### 5️⃣ Construimos cada término de la distancia
Recordemos:
$$D_{ij} = x_i^\top x_i + x_j^\top x_j - 2x_i^\top x_j$$

Ahora hacemos esto en forma matricial.

**Primer término**
Queremos una matriz donde cada fila repita: $x_i^\top x_i$.
Eso se logra con:
$$\boxed{g\mathbf{1}^\top}$$

Dimensiones: $(N \times 1)(1 \times N) = N \times N$
Elemento: $(g\mathbf{1}^\top)_{ij} = g_i = x_i^\top x_i$
✔ correcto

**Segundo término**
Queremos repetir: $x_j^\top x_j$.
Eso se logra con:
$$\boxed{\mathbf{1}g^\top}$$

Dimensiones: $(N \times 1)(1 \times N) = N \times N$
Elemento: $(\mathbf{1}g^\top)_{ij} = g_j = x_j^\top x_j$
✔ correcto

**Tercer término**
Sabemos: $x_i^\top x_j = (XX^\top)_{ij}$.
Entonces:
$$\boxed{2XX^\top}$$

### 6️⃣ Forma matricial final
Sumamos todo:
$$\boxed{D = g\mathbf{1}^\top + \mathbf{1}g^\top - 2XX^\top}$$

Esta es la forma matricial completa que te pide el ejercicio.

### 7️⃣ Verificación rápida (muy importante)
Elemento $(i, j)$:
$$D_{ij} = g_i + g_j - 2(XX^\top)_{ij}$$

Reemplazando:
$$= x_i^\top x_i + x_j^\top x_j - 2x_i^\top x_j$$

✔ coincide exactamente
✔ derivación correcta

### 8️⃣ Interpretación conceptual (esto suma puntos)
Esta expresión muestra que:

$$\boxed{\text{Las distancias euclidianas pueden calcularse usando productos internos}}$$

Esto es la base de:
* ✔ kernels
* ✔ kernel RBF
* ✔ geometría inducida

# Punto 3.2 — Kernel gaussiano (RBF) y geometría inducida

Partimos de la matriz de distancias cuadradas obtenida en 3.1:

$$D_{ij} = \|x_i-x_j\|_2^2$$

A partir de esta matriz se define el kernel gaussiano (RBF) como:

$$K_{ij} = \exp \left( -\frac{\|x_i-x_j\|_2^2}{2\sigma^2} \right)$$

donde:
* $K \in \mathbb{R}^{N \times N}$
* $\sigma > 0$ es el parámetro de ancho del kernel.

### 1️⃣ Por qué los elementos diagonales son constantes
Los elementos diagonales corresponden a:

$$K_{ii} = \exp \left( -\frac{\|x_i-x_i\|_2^2}{2\sigma^2} \right)$$

Pero:
$x_i - x_i = 0$
entonces:
$\|x_i-x_i\|_2^2 = 0$

Sustituyendo:
$$K_{ii} = \exp(0) = 1$$

Por lo tanto:
$$\boxed{K_{ii} = 1 \quad \forall i}$$

✔ Todos los elementos diagonales son constantes
✔ Su valor es 1

### 2️⃣ Interpretación geométrica de la diagonal
Un kernel siempre puede interpretarse como:

$$K_{ij} = \langle \phi(x_i), \phi(x_j) \rangle_{\mathcal H}$$

donde:
* $\phi(x)$ es un mapeo a un espacio de características
* $\mathcal H$ es el RKHS

Entonces:
$$K_{ii} = \langle \phi(x_i), \phi(x_i) \rangle = \|\phi(x_i)\|^2$$

Pero vimos que:
$K_{ii} = 1$

Entonces:
$$\boxed{\|\phi(x_i)\|^2 = 1}$$

**Significado geométrico**
Esto implica que:
$$\boxed{\text{todas las muestras están sobre una esfera de radio 1}}$$
en el espacio de características.

Es decir:
✔ todos los vectores tienen la misma norma
✔ sólo cambian sus direcciones relativas
Esto es una propiedad geométrica fundamental del kernel RBF.

### 3️⃣ Propiedad de simetría
La distancia euclidiana cumple:
$$\|x_i-x_j\|^2 = \|x_j-x_i\|^2$$

Entonces:
$$K_{ij} = \exp \left( -\frac{\|x_i-x_j\|^2}{2\sigma^2} \right) = K_{ji}$$

Por lo tanto:
$$\boxed{K = K^\top}$$

✔ La matriz kernel es simétrica.

### 4️⃣ Propiedad de semidefinición positiva
Un kernel gaussiano es un kernel válido, lo que significa que:
Para cualquier vector:
$c \in \mathbb{R}^N$
se cumple:
$c^\top K c \ge 0$

Por lo tanto:
$$\boxed{K \succeq 0}$$

Es decir:
✔ $K$ es semidefinida positiva

### 5️⃣ Por qué puede interpretarse como matriz de Gram
Una matriz de Gram tiene la forma:
$G_{ij} = \langle v_i, v_j \rangle$
Para algún conjunto de vectores $v_i$.

En kernels:
$$K_{ij} = \langle \phi(x_i), \phi(x_j) \rangle_{\mathcal H}$$

Entonces existe una matriz:
$$\Phi = \begin{bmatrix} \phi(x_1)^\top \\ \phi(x_2)^\top \\ \vdots \\ \phi(x_N)^\top \end{bmatrix}$$

tal que:
$$\boxed{K = \Phi\Phi^\top}$$

Esto es exactamente la forma de una matriz de Gram.